### DISCUSSION TITLE
Install Winget using PowerShell

### DISCUSSION THREAD

**Discussion by whosenbocus** [microsoft/winget-cli#817] [Q&A] [answered]
I'm writing a script that can get any newly built Windows 10 machine to a usable state (all required app installed) using winget. But, since a new Windows 10 installation does not have Winget it sort of make things manual (download and install so that we can use winget). Do we have a way to use PowerShell to install winget?

Thanks,
Wasiim

**Comment by denelon** [comment]
I've seen a gist of a PowerShell script that will fetch the release from the GitHub releases page, but you don't get automatic updates as the store provides on insider builds of Windows today.

**Comment by crutkas** [comment, marked as answer]
Right now i have this Gist i created, it grabs latest off Github.  https://gist.github.com/crutkas/6c2096eae387e544bd05cde246f23901

**Reply by whosenbocus** [reply]
Thanks for the Gist.

Shouldn't the if statement read `if($hasPackageManager)` instead of `if(!$hasPackageManager)`?

If I try this on a fresh win 10 installation with the snippet you provided, it does not run. On the other hand, if I edit the if statement, it runs with the below error. I'm struggling with the dependency part of adding the package.


> Add-AppxPackage : Deployment failed with HRESULT: 0x80073CF3, Package failed updates, dependency or conflict validation.
Windows cannot install package Microsoft.DesktopAppInstaller_1.11.10191.0_x64__8wekyb3d8bbwe because this package depends on a framework that could not be found. Provide the framework 
"Microsoft.VCLibs.140.00.UWPDesktop" published by "CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US", with neutral or x64 processor architecture and minimum 
version 14.0.29016.0, along with this package to install. The frameworks with name "Microsoft.VCLibs.140.00.UWPDesktop" currently installed
Windows cannot install package Microsoft.DesktopAppInstaller_1.11.10191.0_x64__8wekyb3d8bbwe because this package depends on a framework that could not be found. Provide the framework 
"Microsoft.VCLibs.140.00.UWPDesktop" published by "CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US", with neutral or x64 processor architecture and minimum 
version 14.0.29016.0, along with this package to install. The frameworks with name "Microsoft.VCLibs.140.00.UWPDesktop" currently installed are: 
{Microsoft.VCLibs.140.00.UWPDesktop_14.0.27629.0_x64__8wekyb3d8bbwe}
NOTE: For additional information, look for [ActivityId] 6a0cd544-bc46-0001-9017-227f2926d701 in the Event Log or use the command line Get-AppPackageLog -ActivityID 
6a0cd544-bc46-0001-9017-227f2926d701
At line:13 char:3
         Add-AppxPackage -Path $latestRelease.browser_download_url
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     CategoryInfo          : WriteError: (https://github....bbwe.appxbundle:String) [Add-AppxPackage], IOException
     FullyQualifiedErrorId : DeploymentError,Microsoft.Windows.Appx.PackageManager.Commands.AddAppxPackageCommand

**Reply by crutkas** [reply]
This detects if you have the package installed.  If you don't, it goes out to WinGet and gets the appxbundle from https://github.com/microsoft/winget-cli/releases/latest via the API.

The error looks like you're missing Microsoft.VCLibs.140.00.UWPDesktop.  This is actually a good learning as i tested this against a machine that actually already had it installed and missed the depency.  

https://docs.microsoft.com/en-us/troubleshoot/cpp/c-runtime-packages-desktop-bridge#how-to-install-and-update-desktop-framework-packages

**Reply by crutkas** [reply]
Can you try this?

crutkas update: shifted this from appx to msixbundle

```
function InstallWinGet()
{
	$hasPackageManager = Get-AppPackage -name "Microsoft.DesktopAppInstaller"

	if(!$hasPackageManager)
	{
		Add-AppxPackage -Path "https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx"
		
		$releases_url = "https://api.github.com/repos/microsoft/winget-cli/releases/latest"

		[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
		$releases = Invoke-RestMethod -uri "$($releases_url)"
		$latestRelease = $releases.assets | Where { $_.browser_download_url.EndsWith("msixbundle") } | Select -First 1
	
		Add-AppxPackage -Path $latestRelease.browser_download_url
	}
}


```

**Reply by crutkas** [reply]
this assumes x64 but i think this should work

**Reply by whosenbocus** [reply]
Thank you Very Much! Works like a charm!

Yup, running on X46 (Dev Machine).

The IF statement still does not work as intended as running the Get-AppPackage before the initial installation was returning object of Microsoft.DesktopInstaller while the command `winget` was not being recognized by the machine.

I can now add this as part of my script and test it out.

**Reply by michha** [reply]
Script is not working anymore since the format changed from .appxbundle to .msixbundle 4 days ago and the old binaries were deleted.

I refreshed the script from 31st march:
```powershell
$hasPackageManager = Get-AppPackage -name 'Microsoft.DesktopAppInstaller'

if(!$hasPackageManager)
{
    Add-AppxPackage -Path 'https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx'

    $releases_url = 'https://api.github.com/repos/microsoft/winget-cli/releases/latest'

    [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
    $releases = Invoke-RestMethod -uri $releases_url
    $latestRelease = $releases.assets | Where { $_.browser_download_url.EndsWith('msixbundle') } | Select -First 1

    "Installing winget from $($latestRelease.browser_download_url)"
    Add-AppxPackage -Path $latestRelease.browser_download_url
}
```

Works on my machine 😎

**Reply by crutkas** [reply]
yup, this would be the tweak needed, will update my script, thanks @michha!
